### DISCUSSION TITLE
Transparent mode results in "Transparent mode failure: OSError(92, 'Protocol not available')"

### DISCUSSION THREAD

**Discussion by BreakfastSerial** [mitmproxy/mitmproxy#7273] [Q&A] [answered]
#### Problem Description
In transparent mode, mitmproxy seems to fail to proxy requests as expected and the event log prints the following error for each request:

> Transparent mode failure: OSError(92, 'Protocol not available')

#### Steps to reproduce the behavior:
1. Set up fresh netinst Debian, without desktop environment (same issue applies to my TuxedoOS host system)
2. Add entry to `/etc/hosts`  to simulate use-case (e.g. `10.0.0.2 example.com`, 10.0.0.2 being the IP of the mitmproxy system)
3. Download mitmproxy linux binary (https://downloads.mitmproxy.org/11.0.0/mitmproxy-11.0.0-linux-x86_64.tar.gz)
4. Start mitmproxy in transparent mode, listening on port 443: `sudo ./mitmproxy --mode transparent --liste-port 443`
5. Attempt to request https://example.com (`curl -k https://example.com`)
6. View event log of mimtproxy (prints:

```
info: [15:29:22.474] Transparent Proxy listening at *:443.
error: [15:29:23.784] Transparent mode failure: OSError(92, 'Protocol not available')
```

#### System Information
```
./mitmproxy --listen-port 443 --mode transparent --version
Mitmproxy: 11.0.0 binary
Python:    3.12.6
OpenSSL:   OpenSSL 3.3.2 3 Sep 2024
Platform:  Linux-6.1.0-26-amd64-x86_64-with-glibc2.36
```
The same occurs on 

```
./mitmproxy --version
Mitmproxy: 11.0.0 binary
Python:    3.12.6
OpenSSL:   OpenSSL 3.3.2 3 Sep 2024
Platform:  Linux-6.11.0-103009-tuxedo-x86_64-with-glibc2.35
```

**Comment by mhils** [comment]
> 2\. Add entry to `/etc/hosts`  to simulate use-case (e.g. `10.0.0.2 example.com`, 10.0.0.2 being the IP of the mitmproxy system)

This is now how transparent mode works, see https://docs.mitmproxy.org/stable/concepts-modes/#transparent-proxy.

You can start mitmproxy in regular mode and have it deduce the target from the host header.

**Comment by BreakfastSerial** [comment]
My alternative approach was to redirect all outgoing traffic on the target device using nftables: 

```
table ip nat {
        chain output {
                type nat hook output priority filter; policy accept;
                tcp dport 443 dnat to 10.0.0.2:443
        }
}
```

But this results in the same error.  If I redirect  the target domains using hosts entries and don't use the transparent mode, requests/responses show up, but the application runs into an error, which is why I assume transparent mode is the way to go. The host of the target application has the mitmproxy ca trusted. Oddly enough Burp Pro fails completely at intercepting this.

I reckon this issue is more of lack of knowledge about the use of the transparent mode, possibly some weird issue with my (unfortunately NDA closed-source) client application, rather than an transparent mode issue then. Sorry about this and thank you for the swift reply!

**Comment by BreakfastSerial** [comment]
I may be misunderstanding mitmproxy's transparent mode, but I have the current setup in place, to intercept a proxy-agnostic application on system other than what is running mitmproxy. On the target system running the application I have nftables in place:

```
# Set your internal network interface and the mitmproxy port
INTERFACE="wlan0"
PROXY_IP="10.0.0.2"
PROXY_PORT="8080"

# Flush existing rules in the NAT table
nft flush table nat

# Create NAT table and chains
nft add table nat
nft add chain nat prerouting { type nat hook prerouting priority 0 \; }
nft add chain nat output { type nat hook output priority 0 \; }

# DNAT rules to forward HTTP(S) traffic to the remote mitmproxy system
nft add rule nat prerouting iif $INTERFACE tcp dport {80, 443} dnat to $PROXY_IP:$PROXY_PORT
nft add rule nat output tcp dport {80, 443} dnat to $PROXY_IP:$PROXY_PORT
```

And mitmproxy ran into the very same `Transparent mode failure: OSError(92, 'Protocol not available')` upon receiving incoming requests. ChatGPT suggested I should do the following:

```
sudo sysctl -w net.ipv4.ip_forward=1
sudo modprobe nf_conntrack
sudo modprobe nf_nat
sudo sysctl -w net.ipv4.ip_nonlocal_bind=1
```

Which appeared to resolve the issue `Protocol not available`, but now I run into: `Transparent mode failure: FileNotFoundError(2, 'No such file or directory')`.

Oddly enough in this exact setup, if I don't use the transparent mode, I see all requests/responses, but I'm still facing some issues with the functionality of the target application.

If I run mitmproxy in transparent mode as described above, on Kali Linux using the precompiled binary, I get a different result. The event log has numerous occurrences of:

```
info: [08:11:15.930][192.168.0.107:44064] client connect
info: [08:11:15.933][192.168.0.107:44050] server connect 192.168.0.107:8080
```

Until it runs into:

```
error: [08:11:17.874] Unhandled error in task.
Traceback (most recent call last):
  File "asyncio/selector_events.py", line 178, in _accept_connection
  File "socket.py", line 295, in accept
OSError: [Errno 24] Too many open files
```

**Comment by BreakfastSerial** [comment]
I understand the above issue on the Kali's mitmproxy now. I see the target application spams `HEAD google.com/robots.txt` until mitmproxy can't accept new connections anymore. All requests are responded by mitmproxy with a `503 Bad Gateway`. Even manual requests from the target machine (e.g. curl example.com) show up, but are responded with `503 Bad Gateway`.

If I disable transparent mode, I see the requests and responses. Yet I'm quite certain, that transparent mode should be the way to go in my scenario. I'm still trying to find out, why transparent mode results in a 503.

**Comment by BreakfastSerial** [comment, marked as answer]
Lesson learned. This appears to be a specific issue, due to the gRPC communication. If I use httptoolkit to intercept the traffic, using the same redirection methods as described above, the application works as expected.

FWIW I've tried Burp Pro, mitmproxy, caido, ZAP, fiddler, charles, httptoolkit and allproxy. Of all those mitmproxy came the closest, but I believe it cut the gRPC post body messages short, which caused the problems with the application. Httptoolkit was the only one that worked in my case. ZAP did log _some_ requests, but the application did not function correctly.

Just for completeness sake, the application uses tonic/0.12.2 as its gRPC client library.

Sorry about the extensive excursion, deviating from the original issue. I thought it may be good to have my adventure written down, in case someone comes across a similar issue.

**Comment by SurinameClubcard** [comment]
I ran into the same problem. Mitmproxy 12.1.2 on Debian Trixie (Linux kernel 6.12.41). My setup includes a transparent NAT redirection from a downstream Linux system towards mitmproxy. Both port 80 and 443. I got the same error OSError 92 running "mitmproxy --mode transparent". But without "--mode transparent" it works perfectly well.
